### ISSUE TITLE
Diagnosing namespace problems

### ISSUE THREAD

**Issue by flathat** [CORE-POS/IS4C#723]
I'm looking for guidance in resolving:
```
PHP Fatal error:  Class '\COREPOS\pos\lib\models\op\AutoCouponsModel' not found
 in /var/www/IS4C/pos/is4c-nf/install/InstallUtilities.php on line 1105
```
```php
        $models = array(
            '\\COREPOS\\pos\\lib\\models\\op\\AutoCouponsModel',
            '\\COREPOS\\pos\\lib\\models\\op\\CouponCodesModel',
            '\\COREPOS\\pos\\lib\\models\\op\\CustdataModel',
            '\\COREPOS\\pos\\lib\\models\\op\\CustomerNotificationsModel',
            '\\COREPOS\\pos\\lib\\models\\op\\CustPreferencesModel',
            '\\COREPOS\\pos\\lib\\models\\op\\CustReceiptMessageModel',
            '\\COREPOS\\pos\\lib\\models\\op\\CustomReceiptModel',
            '\\COREPOS\\pos\\lib\\models\\op\\DateRestrictModel',
            '\\COREPOS\\pos\\lib\\models\\op\\DepartmentsModel',
            '\\COREPOS\\pos\\lib\\models\\op\\DisableCouponModel',
            '\\COREPOS\\pos\\lib\\models\\op\\DrawerOwnerModel',
            '\\COREPOS\\pos\\lib\\models\\op\\EmployeesModel',
            '\\COREPOS\\pos\\lib\\models\\op\\GlobalValuesModel',
            '\\COREPOS\\pos\\lib\\models\\op\\HouseCouponsModel',
            '\\COREPOS\\pos\\lib\\models\\op\\HouseCouponItemsModel',
            '\\COREPOS\\pos\\lib\\models\\op\\HouseVirtualCouponsModel',
            '\\COREPOS\\pos\\lib\\models\\op\\IgnoredBarcodesModel',
            '\\COREPOS\\pos\\lib\\models\\op\\MasterSuperDeptsModel',
            '\\COREPOS\\pos\\lib\\models\\op\\MemberCardsModel',
            '\\COREPOS\\pos\\lib\\models\\op\\MemtypeModel',
            '\\COREPOS\\pos\\lib\\models\\op\\ParametersModel',
            '\\COREPOS\\pos\\lib\\models\\op\\ProductsModel',
            '\\COREPOS\\pos\\lib\\models\\op\\ShrinkReasonsModel',
            '\\COREPOS\\pos\\lib\\models\\op\\SpecialDeptMapModel',
            '\\COREPOS\\pos\\lib\\models\\op\\SubDeptsModel',
            '\\COREPOS\\pos\\lib\\models\\op\\TendersModel',
            '\\COREPOS\\pos\\lib\\models\\op\\UnpaidArTodayModel',
            // depends on custdata
            '\\COREPOS\\pos\\lib\\models\\op\\MemberCardsViewModel',
        );
        foreach ($models as $class) {
            $obj = new $class($db);  // L#1105
            $errors[] = $obj->createIfNeeded($name);
        }
```
when executing `$LANE/install/index.php` (Necessities), after "Testing operational DB Connection:"

The classfile exists and contains the class, begins:
```php
namespace COREPOS\pos\lib\models\op;
use COREPOS\pos\lib\models\BasicModel;

/**
  @class AutoCouponsModel
*/
class AutoCouponsModel extends BasicModel
```
`InstallUtilities.php` seems to be executing the `AutoLoader` and other programs in the install suite seem to be able to use classes, e.g. `ParametersModel` from that set.  I don't understand what L#1105 is expecting that didn't happen.

**Comment by gohanman** [comment]
It's expecting the autoloader to locate that file. Looking through the commit history on the autoloader I don't see anything related to mapping namespaces to filesystem paths that has changed though. You'd have to add some manual logging to https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/lib/AutoLoader.php#L73-L88 to figure what's going awry.

**Comment by flathat** [comment]
On another lane (lane-on-server, like the one reporting the error) that is configured the same way as far as I know the autoloader does locate that file. I wonder if something somewhere else has gone wrong. Should these classes all already be in `$map`?

On the lane in question this is the branch at L#73, which I assume should be taken for the members of `$models`, e.g. `'\\COREPOS\\pos\\lib\\models\\op\\AutoCouponsModel'` that are not already in `$map` but I don't see how it can be because `strpos($name, '\\') == 0`
```php
        } elseif (!isset($map[$name]) && strpos($name, '\\') > 0) {
            $pieces = explode('\\', $name);
            $s = DIRECTORY_SEPARATOR;
            if (count($pieces) > 2 && $pieces[0] == 'COREPOS' && $pieces[1] == 'common') {
                $path = dirname(__FILE__) . $s . '..' . $s . '..' . $s . '..' . $s . 'common' . $s;
                $path .= self::arrayToPath(array_slice($pieces, 2));
                if (file_exists($path)) {
                    $map[$name] = $path;
                }

            } elseif (count($pieces) > 2 && $pieces[0] == 'COREPOS' && $pieces[1] == 'pos') {
                $path = dirname(__FILE__) . $s . '..' . $s;
                $path .= self::arrayToPath(array_slice($pieces, 2));
                if (file_exists($path)) {
                    $map[$name] = $path;
                }
            }
```
If I add another branch testing `strpos($name, '\\') == 0` the page finishes and no error is reported:
```php
        } elseif (!isset($map[$name]) && strpos($name, '\\') == 0) {
            $pieces = explode('\\', $name);
            // NOTE array_shift()
            $x = array_shift($pieces);
            $s = DIRECTORY_SEPARATOR;
            if (count($pieces) > 2 && $pieces[0] == 'COREPOS' && $pieces[1] == 'common') {
                $path = dirname(__FILE__) . $s . '..' . $s . '..' . $s . '..' . $s . 'common' . $s;
                $path .= self::arrayToPath(array_slice($pieces, 2));
                if (file_exists($path)) {
                    $map[$name] = $path;
                }
            } elseif (count($pieces) > 2 && $pieces[0] == 'COREPOS' && $pieces[1] == 'pos') {
                $path = dirname(__FILE__) . $s . '..' . $s;
                $path .= self::arrayToPath(array_slice($pieces, 2));
                if (file_exists($path)) {
                    $map[$name] = $path;
                }
            }
        }
```

**Comment by gohanman** [comment]
There shouldn't be a backslash at position zero in the autoloader. https://repl.it/qOH/269

This may have been a change in behavior introduced in PHP 5.3.3 based on some stackoverflow conversations and common minimum requirements of many projects although PHP's release notes aren't explicit about it.

**Comment by flathat** [comment]
I've added the `strpos($name, '\\') == 0` branch to my version of `AutoLoader.php` in order to make it work but can make no sense of what I see compared to my other lane where this isn't needed.

**Comment by gohanman** [comment]
It would be a much better idea to use >= 0. There's no guarantee a
namespaced class name is going to include a leading backslash at position
0.

**Comment by gohanman** [comment]
I found [the bug](https://bugs.php.net/bug.php?id=50731) that was fixed in PHP 5.3.3.
